state/processor.go

```diff
@@ -131,33 +136,49 @@ func (tp *TransactionProcessor) ApplyTransactions(layer types.LayerID, txs []*ty
 		remainingCount = len(remaining)
 	}
 
-	newHash, err := tp.globalState.Commit(false)
+	newHash, err := tp.Commit(false)
 
 	if err != nil {
 		return remainingCount, fmt.Errorf("failed to commit global state: %v", err)
 	}
 
-	tp.addStateToHistory(layer, newHash)
+	tp.Log.Info("new state root for layer %v is %x", layer, newHash)
+	tp.Log.With().Info("new state", log.Uint64("mesh.LayerID", uint64(layer)), log.String("root_hash", newHash.String()))
 
-	return remainingCount, nil
+	err = tp.addStateToHistory(layer, newHash)
+
+	return remainingCount, err
 }
 
-func (tp *TransactionProcessor) addStateToHistory(layer types.LayerID, newHash types.Hash32) {
-	tp.stateQueue.PushBack(newHash)
-	if tp.stateQueue.Len() > maxPastStates {
-		hash := tp.stateQueue.Remove(tp.stateQueue.Front())
-		tp.db.Commit(hash.(types.Hash32), false)
+func (tp *TransactionProcessor) addStateToHistory(layer types.LayerID, newHash types.Hash32) error {
+	tp.trie.Reference(newHash, types.Hash32{})
+	err := tp.trie.Commit(newHash, false)
+	if err != nil {
+		return err
 	}
-	tp.prevStates[layer] = newHash
-	tp.db.Reference(newHash, types.Hash32{})
-	tp.With().Info("new state root added to history",
-		log.LayerId(layer.Uint64()),
-		log.String("state_root", util.Bytes2Hex(newHash.Bytes())),
-	)
+	err = tp.addState(newHash, layer)
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
-func (tp *TransactionProcessor) GetStateRoot() types.Hash32 {
-	return tp.stateQueue.Back().Value.(types.Hash32)
+func getStateRootLayerKey(layer types.LayerID) []byte {
+	return []byte(NewRootKey + strconv.FormatUint(uint64(layer), 10))
```

Comment:
```suggestion
	return append([]byte(NewRootKey), layer.ToBytes()...)
```
Or even better - store `NewRootKey` as bytes and then there's no need for a conversion to bytes at all.